python/tk_framework_desktopserver/shotgun/api_v2.py

```diff
@@ -0,0 +1,780 @@
+                        err="",
+                        out="",
+                    ),
+                )
+                return
+
+        project_entity, entities = self._get_entities_from_payload(data)
+
+        # It shouldn't ever happen at this stage of the workflow, but we'll
+        # check to make sure that we can get an environment for the type of
+        # we're dealing with here.
+        env_name = self.__pick_environment(project_entity, entities[0])
+
+        if env_name is None:
+            self.logger.error(
+                "The provided entity does not correspond to any environment."
+            )
+            return
+
+        args_file = self._get_arguments_file(
+            dict(
+                config=data["pc"],
+                name=data["name"],
+                entities=entities,
+                project=project_entity,
+                sys_path=sys.path,
+                base_configuration=self.BASE_CONFIG_URI,
+                engine_name=self.ENGINE_NAME,
+            ),
+        )
+
+        script = os.path.join(
+            os.path.dirname(__file__),
+            "scripts",
+            "execute_command.py"
+        )
+
+        output = None
+
+        try:
+            output = subprocess.check_output(
+                [sys.executable, script, args_file]
+            )
+        except subprocess.CalledProcessError:
+            if output:
+                self.logger.error(output)
+            raise
+        else:
+            self.logger.debug(output)
+
+        self.logger.debug("Command execution complete.")
+        self.host.reply(
+            dict(
+                retcode=self.COMMAND_SUCCEEDED,
+                out="Command executed successfully.",
+                err="",
+            ),
+        )
+
+    def get_actions(self, data):
+        """
+        RPC method that sends back a dictionary containing engine commands
+        for each pipeline configuration associated with the project.
+
+        :param dict data: The data passed down by the client. At a minimum,
+            the dict should contain the following keys: project_id, entity_id,
+            entity_type, pipeline_configs, and user.
+        """
+        # If we weren't sent a usable entity id, we can just query the first one
+        # from the project. This isn't a big deal for us, because we're only
+        # concerned with picking the correct environment when we bootstrap
+        # during a caching operation. This is going to be the situation when
+        # a page is pre-loading entity-level commands on page load, as opposed to
+        # passing down a specific entity that's been selected in an already-loaded
+        # page.
+        if "entity_id" in data and data["entity_id"] == -1:
+            if data["entity_type"] == "Project":
+                data["entity_id"] = data["project_id"]
+            else:
+                temp_entity = self._engine.shotgun.find_one(
+                    data["entity_type"],
```

Comment:
will this be an issue with non-project entities? What if we wanted actions on the software entity? or the project entity?